File: class16/java/PES1601-FluxoPedido.java
/*
 * PES1601 - Implementação do Fluxo Modelado (Pedido)
 * Aula 16: Diagramas de Sequência e de Atividades
 * Projeto e Engenharia de Software - Prof. Giovane Barcelos
 *
 * Implementa o fluxo de pedido modelado nos diagramas de sequência
 * e atividades, demonstrando a tradução de UML para código.
 *
 * Compilar/executar:
 *   javac PES1601-FluxoPedido.java && java FluxoPedido
 */

import java.util.*;

class EstoqueService {
    private Map<String, Integer> estoque = new HashMap<>();

    EstoqueService() {
        estoque.put("Produto A", 10);
        estoque.put("Produto B", 5);
        estoque.put("Produto C", 0);
    }

    boolean verificarDisponibilidade(List<String> itens, List<String> indisponiveis) {
        boolean tudoDisponivel = true;
        for (String item : itens) {
            if (estoque.getOrDefault(item, 0) <= 0) {
                indisponiveis.add(item);
                tudoDisponivel = false;
            }
        }
        return tudoDisponivel;
    }

    void reservar(List<String> itens) {
        for (String item : itens) estoque.merge(item, -1, Integer::sum);
    }

    void liberar(List<String> itens) {
        for (String item : itens) estoque.merge(item, 1, Integer::sum);
    }
}

class PagamentoService {
    static boolean processar(double valor) {
        return valor <= 10000;
    }
}

class PedidoService {
    private EstoqueService estoque = new EstoqueService();

    boolean criarPedido(String cliente, List<String> itens, double valor) {
        System.out.printf("%n%s%n", "=".repeat(50));
        System.out.printf("  PEDIDO — %s%n", cliente);
        System.out.printf("%s%n", "=".repeat(50));

        List<String> indisponiveis = new ArrayList<>();
        if (!estoque.verificarDisponibilidade(itens, indisponiveis)) {
            System.out.printf("  ✗ Itens indisponíveis: %s%n", String.join(", ", indisponiveis));
            return false;
        }

        estoque.reservar(itens);
        System.out.printf("  ✓ Itens reservados: %s%n", String.join(", ", itens));

        if (!PagamentoService.processar(valor)) {
            estoque.liberar(itens);
            System.out.printf("  ✗ Pagamento recusado (valor: R$ %.2f)%n", valor);
            return false;
        }
        System.out.printf("  ✓ Pagamento aprovado: R$ %.2f%n", valor);

        System.out.printf("  ✓ Pedido confirmado!%n");
        System.out.printf("  ✓ Email de confirmação enviado para %s%n", cliente);
        return true;
    }
}

class FluxoPedido {
    public static void main(String[] args) {
        PedidoService ps = new PedidoService();

        ps.criarPedido("ana@email.com", Arrays.asList("Produto A", "Produto B"), 150.00);
        ps.criarPedido("carlos@email.com", Arrays.asList("Produto C"), 50.00);
        ps.criarPedido("maria@email.com", Arrays.asList("Produto A"), 15000.00);

        System.out.printf("%n✓ O código reflete fielmente o fluxo dos diagramas UML.%n");
    }
}

File: class16/python/PES1601-FluxoPedido.py
"""
PES1601 - Implementação do Fluxo Modelado (Pedido)
Aula 16: Diagramas de Sequência e de Atividades
Projeto e Engenharia de Software - Prof. Giovane Barcelos

Implementa o fluxo de pedido modelado nos diagramas de sequência
e atividades, demonstrando a tradução de UML para código.
"""


class EstoqueService:
    def __init__(self):
        self._estoque = {"Produto A": 10, "Produto B": 5, "Produto C": 0}

    def verificar_disponibilidade(self, itens):
        indisponiveis = [item for item in itens
                         if self._estoque.get(item, 0) <= 0]
        return len(indisponiveis) == 0, indisponiveis

    def reservar(self, itens):
        for item in itens:
            self._estoque[item] -= 1

    def liberar(self, itens):
        for item in itens:
            self._estoque[item] += 1


class PagamentoService:
    @staticmethod
    def processar(valor):
        # Simulação: pagamentos acima de R$ 10.000 falham (regra de negócio)
        return valor <= 10000


class PedidoService:
    def __init__(self):
        self.estoque = EstoqueService()

    def criar_pedido(self, cliente, itens, valor):
        print(f"\n{'='*50}")
        print(f"  PEDIDO — {cliente}")
        print(f"{'='*50}")

        # Passo 1: verificar disponibilidade
        disponivel, indisponiveis = self.estoque.verificar_disponibilidade(itens)
        if not disponivel:
            print(f"  ✗ Itens indisponíveis: {', '.join(indisponiveis)}")
            return False

        # Passo 2: reservar
        self.estoque.reservar(itens)
        print(f"  ✓ Itens reservados: {', '.join(itens)}")

        # Passo 3: processar pagamento
        if not PagamentoService.processar(valor):
            self.estoque.liberar(itens)
            print(f"  ✗ Pagamento recusado (valor: R$ {valor:.2f})")
            return False
        print(f"  ✓ Pagamento aprovado: R$ {valor:.2f}")

        # Passo 4: confirmar
        print(f"  ✓ Pedido confirmado!")
        print(f"  ✓ Email de confirmação enviado para {cliente}")
        return True


if __name__ == "__main__":
    ps = PedidoService()

    # Cenário 1: sucesso
    ps.criar_pedido("ana@email.com", ["Produto A", "Produto B"], 150.00)

    # Cenário 2: item indisponível
    ps.criar_pedido("carlos@email.com", ["Produto C"], 50.00)

    # Cenário 3: pagamento recusado
    ps.criar_pedido("maria@email.com", ["Produto A"], 15000.00)

    print(f"\n✓ O código reflete fielmente o fluxo dos diagramas UML.")
